teps/0048-task-results-without-results.md

```diff
@@ -0,0 +1,425 @@
+`finally` clause throws error and is not executed.
+
+### Option D
+
+`Pipeline` author can check the status of the result producing task and design the `pipeline` to handle success and
+failure scenarios.
+
+Pros/Cons:
+
++ `Pipeline` author has all the freedom to design the `pipeline`.
++ Can be extended to `continueAfterSkip` with missing task results (provided execution status is available within `dag`
+tasks)
+- Cannot work with `when` expressions
+- If `pipeline` author does not design failure scenario, what value should be used for the `param` referring to task
+results?
+
+### Option E
+
+The `cleanup` is necessary when the resources are acquired but not needed if the acquisition fails. This option is
+proposing introducing `when` expression to `finally`. `Cleanup` task in `finally` checks the execution status of
+`acquire` task, and guards `cleanup` if `acquire` failed. This proposal helps avoid dealing with missing task results
+in `finally` i.e. no default or no zero values for task results needed.
+
+Pros/Cons:
+
++ Introducing guarded finally task, its a clean design.
++ `Pipeline` author has full control over its design.
++ Can be extended to `continueAfterSkip`
+- no defaults needed
+- Cannot work with `when` expressions
+
+### Option F
+
+Allow specifying `default` with the param in the `pipelineTask` where a parameter has a reference to task result.
+
+```yaml
+spec:
+  params:
+    - name: owner-name
+      default: "the-best-owner"
+  tasks:
+   - name: boskos-acquire
+     taskRef:
+       name: boskos-acquire
+     params:
+       - name: owner-name
+         value: $(params.owner-name)
+   - name: do-stuff-with-resource
+     taskRef:
+       name: do-stuff-with-resource
+     params:
+       - name: resource-name
+         value:  $(tasks.boskos-acquire.results.leased-resource)
+ finally:
+   - name: boskos-release
+     taskRef:
+       name: boskos-release
+     params:
+       - name: leased-resource
+         value:  $(tasks.boskos-acquire.results.leased-resource)
+         default: "fake-awesome-project"
+```
+
+Pros/Cons:
+
++ This alternative allows task result consumer to decide the default value based on its use case v/s task
+  result producer defining common default for all consumers.
+- Not compatible with `when` expression.
+
+**Analysis:** We have listed `when` expression with missing task results as one of the possible use case and something
+needs to be addressed with the proposal in this TEP. But having difficulty coming up with a possibility of getting into
+`when` expression with missing task results. Let's take an example for the same,
+A(guarded)(continueAfterSkip=true) -> B(guarded with task results from A)  -> C
+This example shows that `B` is guarded with task results from `A` where `A` was skipped but is it really a possible
+scenario? When this `pipeline` says continue executing `B` even though `A` was skipped, why would `B` be guarded with a
+task result from `A`?
+
```

Comment:
Another proposal that came up (I'm not sure where?? i feel like @chhsia0 was involved but not sure) is to introduce a defaulting syntax when using the variable interpolation for results, e.g. something like:

`$(tasks.foo.results.bar?somedefault)` (or maybe instead of somedefault we require reference to another var?)

I might be kinda crazy but i actually like this option b/c it puts the control completely on the consuming PipelineTask (like option b) but also works for when expressions

if we go this route hopefully we could borrow the syntax from somewhere else

Reply:
this is interesting -- just wondering how it'd look like if the default was a variable 

would it be `$(tasks.foo.results.bar?$(params.someDefault))` 🙃  -- maybe `$(tasks.foo.results.bar?params.someDefault)` is better?

but then how would this look like if we add support for array results proposed in https://github.com/tektoncd/community/pull/477?

(back to focusing on the problem statement only lol)